<a href="https://colab.research.google.com/github/OlgaZy/Python_for_DA/blob/main/HW_15_3_%D0%A2%D0%B5%D1%81%D1%82%D0%B8_%D0%B4%D0%BB%D1%8F_%D0%BC%D0%B0%D0%BB%D0%B8%D1%85_%D0%B2%D0%B8%D0%B1%D1%96%D1%80%D0%BE%D0%BA_%D1%82%D0%B0_%D0%BF%D1%80%D0%BE%D0%BF%D0%BE%D1%80%D1%86%D1%96%D0%B9.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Тести для малих вибірок та пропорцій



In [ ]:
import numpy as np
from scipy import stats

## **Завдання 1**
E-commerce компанія після редизайну сайту підозрює, що **середній час до покупки (time-to-purchase)** користувачів **збільшився**.
Історично середній час від першого заходу на сайт до покупки становив $\mu_0 = 123.7$ хвилин.

Команда аналітиків випадково обрала дані **7 покупців після редизайну**:
`128, 135, 121, 142, 126, 151, 123`

З допомогою наявних даних зʼясуйте, чи збільшився середній час покупки після редизайну.

Для цього виконайте наступні 6 кроків. Правильне виконання кожного з кроків оцінюється в 1 бал.

1. Запишіть нульову та альтернативну гіпотези і визначте тип тесту.
2. Обчисліть вибіркові статистики: $\bar x$, $s$, $n$.
3. Оберіть тип тесту та виконайте його (знайдіть критичне значення тесту, статистику тесту та р-значення) будь-яким пасуючим способом, щоб перевірити гіпотезу на рівні значущості $\alpha = 0.10$.
4. Прийміть рішення, чи відхиляєте ви гіпотезу $H_0$ використовуючи p-value.
5. Напишіть висновок: чи справді редизайн сайту подовжив час до покупки?
6. Чи зміниться ваше рішення при зміні рівня значущості на $\alpha = 0.05$.


**Крок 1: Формулюємо гіпотези**

H₀: μ = 123.7  хв (середній час до покупки (time-to-purchase) становить 123.7  хв)

H₁: μ > 123.7  хв (після реизайну середній час до покупки (time-to-purchase) став більшим за 123.7  хв)

Тип тесту: правосторонній (one-tailed)

Рівень значущості  𝛼=0.10

**Крок 2. Обчислення вибіркової статистики: $\bar x$, $s$, $n$.**

**Крок 3. Вибір типу тесту та виконання його (критичне значення тесту, статистики тесту та р-значення), щоб перевірити гіпотезу на рівні значущості  𝛼=0.10.**

In [ ]:
# Дані з умови
time_to_purchase = np.array([128, 135, 121, 142, 126, 151, 123])
mu0 = 123.7   # середній час до покупки
alpha = 0.10

# 1. Обчислимо вибіркову статистику
x_bar = np.mean(time_to_purchase)
s = np.std(time_to_purchase, ddof=1)   # стандартне відхилення з n-1
n = len(time_to_purchase)

# 2. Обчислимо t-статистику вручну
t_stat = (x_bar - mu0) / (s / np.sqrt(n))

# 3. Критичне значення для правостороннього тесту
df = n - 1
t_crit = stats.t.ppf(1-alpha, df)

#  4. p-value
p_value = 1- stats.t.cdf(t_stat, df)

**Крок 4. Прийміть рішення, чи відхиляєте ви гіпотезу  $H_0$  використовуючи p-value.**

In [ ]:
print(f"Вибіркове середнє: {x_bar:.3f}")
print(f"Вибіркове стандартне відхилення: {s:.3f}")
print(f"t-статистика: {t_stat:.3f}")
print(f"Критичне значення (df={df}): {t_crit:.3f}")
print(f"p-value: {p_value:.3f}")

# Крок 5. Висновок
if t_stat > t_crit:
    print(f"Відхиляємо H0: середній час до покупки (time-to-purchase) більше {mu0} хв на рівні значущості {alpha}")
else:
    print("Не відхиляємо H0: доказів недостатньо.")

Вибіркове середнє: 132.286
Вибіркове стандартне відхилення: 10.981
t-статистика: 2.069
Критичне значення (df=6): 1.440
p-value: 0.042
Відхиляємо H0: середній час до покупки (time-to-purchase) більше 123.7 хв на рівні значущості 0.1


**Крок 5. Використовуючи лише 1 метод з бібліотеки scipy, який одразу вертає і статистику тесту, і р-значення.**

In [ ]:
res = stats.ttest_1samp(time_to_purchase, popmean=mu0, alternative="greater")


t_stat_by_scipy = round(res.statistic, 3)
p_value_by_scipy = round(res.pvalue, 3)
t_crit = round(stats.t.ppf(1-alpha, res.df), 3)

print(f"t-статистика = {t_stat_by_scipy:}, p-value = {p_value_by_scipy:}")
print(f"alpha = {alpha:}")
print(f"t_crit = {t_crit:}")

if p_value_by_scipy < alpha:
    print(f"Відхиляємо H0: середній час до покупки більше {mu0} хв на рівні значущості {alpha}")
else:
    print(f"Не відхиляємо H0: доказів недостатньо на рівні значущості {alpha}")


if t_stat > t_crit:
    print(f"Відхиляємо H0: середній час до покупки більше {mu0} хв на рівні значущості {alpha}")
else:
    print("Не відхиляємо H0: доказів недостатньо")


t-статистика = 2.069, p-value = 0.042
alpha = 0.1
t_crit = 1.44
Відхиляємо H0: середній час до покупки більше 123.7 хв на рівні значущості 0.1
Відхиляємо H0: середній час до покупки більше 123.7 хв на рівні значущості 0.1


**Крок 6. Чи зміниться ваше рішення при зміні рівня значущості на  𝛼=0.05 .**

In [ ]:
alpha = 0.05

if p_value_by_scipy < alpha:
    print(f"Відхиляємо H0: середній час до покупки більше {mu0} хв на рівні значущості {alpha}")
else:
    print(f"Не відхиляємо H0: доказів недостатньо на рівні значущості {alpha}")


Відхиляємо H0: середній час до покупки більше 123.7 хв на рівні значущості 0.05


**Висновок:**
Редизайн сайту подовжив час до покупки (середній час став більшим) на рівні значущості 𝛼= 0.1 , а також для 𝛼=0.05 рішення незмінне.

## **Завдання 2.**

До спеціальної рекламної кампанії **23%** дорослих упізнавали логотип компанії. Після завершення кампанії відділ маркетингу провів опитування: з **1200** випадково відібраних дорослих **311** упізнали логотип.

Перевірте на рівні значущості **3%** ($\alpha=0.03$), чи дають ці дані достатні підстави стверджувати, що **тепер більше ніж 23%** дорослих упізнають логотип компанії. Для розвʼязку використовуйте бібліотеку `statsmodels`.

Зробіть висновок, чи зросла впізнаваність логотипу.

Додатково, обчисліть довірчий інтревал на заданому рівні значущості і проінтерпретуйте текстом - як він додатково пояснює прийняте нами рішення?

In [20]:
import statsmodels.api as sm
from statsmodels.stats.proportion import proportions_ztest, proportion_confint

count = 311   # кількість "успіхів"
nobs = 1200    # обсяг вибірки
p0 = 0.23      # гіпотетична пропорція (H0)
alpha = 0.03

# z-тест для пропорції
prop_stat, prop_pval = proportions_ztest(count, nobs, p0, alternative='larger')
prop_confint = proportion_confint(count, nobs, alpha=0.03, method='normal')

print("🔹 Тест для однієї пропорції")
print(f"Z-статистика = {prop_stat:.3f}")
print(f"p-value = {prop_pval:.4f}")
print(f"97% довірчий інтервал: {prop_confint}")

🔹 Тест для однієї пропорції
Z-статистика = 2.306
p-value = 0.0106
97% довірчий інтервал: (0.23171700302179205, 0.28661633031154127)


In [21]:
if prop_pval < alpha:
    print(f"Відхиляємо H0: Є достатні підстави стверджувати, що тепер більше ніж {p0*100}% дорослих упізнають логотип компанії.")
else:
    print(f"Не відхиляємо H0: доказів недостатньо на рівні значущості {alpha}.")

Відхиляємо H0: Є достатні підстави стверджувати, що тепер більше ніж 23.0% дорослих упізнають логотип компанії.


**Висновок:**

Оскільки p-значення (0.0106) менше за рівень значущості $\alpha=0.03$, ми відхиляємо нульову гіпотезу. Є достатні підстави стверджувати, що після рекламної кампанії більше ніж 23% дорослих упізнають логотип компанії.

Довірчий інтервал (0.232, 0.287) показує, що справжня частка дорослих, які впізнають логотип, з імовірністю 97% знаходиться в цьому діапазоні. Оскільки нижня межа (23.6%) перевищує 23%, це узгоджується з результатом тесту й додатково підтверджує, що впізнаваність логотипу справді зросла.